src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,94 +10,183 @@ internal struct FileStatus
+                {
+                    // User belongs to a group that effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRGRP;
+                    writeBit = Interop.Sys.Permissions.S_IWGRP;
+                }
+                else
+                {
+                    // Others permissions
+                    readBit = Interop.Sys.Permissions.S_IROTH;
+                    writeBit = Interop.Sys.Permissions.S_IWOTH;
+                }
+#endif
+
+                return (_fileCache.Mode & (int)readBit)  != 0 && // has read permission
+                       (_fileCache.Mode & (int)writeBit) == 0;   // but not write permission
             }
-            else
+        }
+
+        // Checks if the main path is a symbolic link
+        // Only call if Refresh has been successfully called at least once
+        private bool HasSymbolicLinkFlag
+        {
+            get
             {
-                // Others permissions
-                readBit = Interop.Sys.Permissions.S_IROTH;
-                writeBit = Interop.Sys.Permissions.S_IWOTH;
+                Debug.Assert(IsFileCacheInitialized);
+                return (_fileCache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
             }
-#endif
+        }
+
+        // Checks if the specified cache (lstat=_fileCache, stat=_symlinkCache) has the directory attribute set
+        // Only call if Refresh has been successfully called at least once, and you're
+        // certain the passed-in cache was successfully retrieved
+        private bool CacheHasDirectoryFlag(Interop.Sys.FileStatus cache) => (cache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
+
+        internal static void Initialize(ref FileStatus status, bool initiallyDirectory)
+        {
+            status.InitiallyDirectory = initiallyDirectory;
+            status.InvalidateCaches();
+        }
+
+        // Sets the cache initialization flags to -1, which means the caches are now uninitialized
+        internal void InvalidateCaches()
+        {
+            _initializedFileCache = -1;
+            _initializedSymlinkCache = -1;
+        }
 
-            return ((_fileStatus.Mode & (int)readBit) != 0 && // has read permission
-                (_fileStatus.Mode & (int)writeBit) == 0);     // but not write permission
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return HasReadOnlyFlag;
         }
 
-        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        internal bool IsHidden(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName, bool continueOnError = false)
         {
-            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
+            EnsureCachesInitialized(path, continueOnError);
+            return IsNameHidden(fileName) || HasHiddenFlag;
```

Comment:
This saves one native call in `.*` cases.
```suggestion
            if (IsNameHidden(fileName))
            {
                return true;
            }
            
            EnsureCachesInitialized(path, continueOnError);
            return HasHiddenFlag;
```